Automation Decorator for custom components

rmkraus opened the pull request:
I gave serious consideration to calling this PR, "Features that make @balloob cringe", but decided better of it at the last minute. :stuck_out_tongue_winking_eye: 

What I have done here is created an Automation function decorator. This decorator is only to be used when creating custom components. It allows Python scripts to be written in, what I am suggesting is, a more concise way. I have an example of the decorator in use [here](https://gist.github.com/rmkraus/654fe4728cad7d125c9e).

In the example, both _mute_light.py_ and _mute_light_old.py_ will turn on a Hue light to red while the TV is muted and then off when mute is turned off. This is obviously logic that would be better implemented using the automation component, but for example's sake, I have put it into a Python component. _mute_light_old.py_ is the current way of doing this. _mute_light.py_ is the way this would be done using this decorator.

Decorated functions contain the standard automation declaration in yaml format in their docstrings (minus the action key). The same logic used to setup and run automations is then used to setup and run the decorated function. The setup function in the component calls _automation.activate_ with the domain name for the component. This activates all automations under the specified domain.

The following are the potential benefits to this:
1. Concise automation declarations. The triggers and the logic are defined under the same function. In the old format, the triggers are defined in setup and the actions are defined in a different function. Logically, these should be together.
2. Allows for very Pythonic layout. When compared to [custom_components/example.py](https://github.com/balloob/home-assistant/blob/dev/config/custom_components/example.py), the decorated layout allows different pieces of logic to be broken up into wholly different functions. For complex logic and many custom automations, the decorated syntax doesn't require defining many functions inside of one. It also doesn't require that all the logic be in one file, it may be broken up among multiple files inside the same module.
3. Because of the very sequential nature of the function definitions, this new format would easily lend itself being programmed with a visual programming language like Blockly. Just saying.....
4. Consistent formatting with standard automation. For new users moving from the default automation component to custom components, the syntax is familiar and easy to learn.
5. Automations are automations are automations. Whether the logic is programmed in Python or in YAML, they are setup and handled by the automation component.
6. Automation functions have consistent arguments. No matter how the function is set to be triggered, it will always have the same argument.
7. Automation rules have immediate access to information about the time it last completed, the number of times it is currently running (if it were to be triggered twice at once). Just the same, automation rules can access this information about other automation rules as well.
8. Python Automation rules put information in the log, just like YAML Automation rules.
9. No globals required. Writing custom components without using the global keyword becomes very difficult as logic complexity increases.

The following are the potential drawbacks:
1. Multiple ways to do the same thing. This decorator certainly isn't strictly necessary. It is merely a helper for writing custom components. I feel this is ok as users do not appear to be commonly writing custom components currently. Making it easier should increase the number of users willing to write custom components rather than requesting more complex features in the automation and script components.
2. MVP means Maximum Viable Product... right?

This is not pull ready, but I wanted to get some feedback on it before putting in any more work. I think it is neat and fills a very real need.

rmkraus committed:
Created automation decorator prototype

Created an initial iteration of an Automation decorator.

balloob commented:
![ojxnygv](https://cloud.githubusercontent.com/assets/1444314/12533988/7ee0403a-c1fa-11e5-9624-5efa3bcddaa7.gif)

maddox commented:
![](http://www.reactiongifs.com/wp-content/uploads/2011/05/MichealJacksonPopcorn.gif)

rmkraus commented:
![You're Welcome](http://i.giphy.com/xag8xIt7P4fGE.gif)

rmkraus committed:
Reverting Automation decorator in favor of a new approach.

rmkraus committed:
Moved sunrise/sunset tracking to helpers

The automation component contained some very handy and generic
functions for tracking sunset and sunrise. This was moved to
helpers/event.py.

rmkraus committed:
Initial pass at event decorators

Created event decorators for custom components. Decorators were created
for the events: track_state_change, track_sunrise, track_sunset, and
track_time_change.

rmkraus commented:
After talking to Paulus about this idea a bit, we came up with a cleaner way of doing this. This PR now creates decorators like

```python
@track_state_change(ENTITY_IDS, [FROM], [TO])
def ...

@track_sunrise([OFFSET])
def ...

@track_sunset([OFFSET])
def ...

@track_time_change([YEAR], [MONTH], [DAY], [HOUR], [MINUTE], [SECOND])
def ...
```

All decorators are activated by bootstrap after loading the components. 

Functions wrapped by these decorators will still have the additional properties: alias, domain, is_running, num_running, activated, last_run. Many of these are very helpful for complex logic. 

A logbook entry will still be created each time the function is triggered.

No testing has been done yet, I just wanted to update the idea with Paulus's suggestions. I will now do some testing of this, update my example, and consider the effects of chaining these decorators together (because someone is going to do it).

rmkraus committed:
Fixed bugs to allow HA to boot again

1) helpers/event should not import the sun component unless it is
requested. This prevents circular import.
2) fixed import typo in bootstrap
2) bootstrap cannot import event_decorators until it is needed because
this leads to a circular import.

rmkraus committed:
A few lint fixes to event decorators.

rmkraus committed:
Another revision on event decorators

This revision of event decorators removes much of the complexity. The
decorated functions are no longer wrapped with a class that tracks
last_run, etc. Bootstrap now gives hass to the event_decorators module
before initializing components so the decorators no longer require
activation.

balloob commented:
This looks great, I like it :sparkles: . It should include a few tests to make sure it works.

rmkraus commented:
![](http://www.reactiongifs.com/r/jsc.gif)

rmkraus commented:
[The example is updated now.](https://gist.github.com/rmkraus/654fe4728cad7d125c9e)

rmkraus committed:
Many updates regarding event decorators

1. Added HASS to the arguments for callbacks that are created with
event decorators.
2. Added a service decorator.
3. Updated example.py in the example config to use the event decorators.

rmkraus commented:
HASS is now the first argument passed to the callbacks that are configured with the decorators. The example.py has also been updated to use the decorators. Tests are on the way.

rmkraus committed:
Updated example.py component

Cleaned up example.py to better handle failed loads.

balloob commented on homeassistant/helpers/event_decorators.py:
```diff
@@ -0,0 +1,78 @@
+    action(HASS, *args, **kwargs)
+
+
+def service(domain, service):
```
This should not be an event decorator but be in `helpers/service.py`

rmkraus committed:
Lint fixes and additions to event decorators

1. service decorator was overwriting the function name with one of its
arguments.
2. Accidentally left an extra argument in track_sunrise.
3. Added track_utc_time_change decorator.

rmkraus committed:
Added event helper tests

1. Added tests for all event decorators
2. Added tests for sunrise and sunset event helpers

rmkraus committed:
Moved service decorator to service helpers

Moved the service decorator to the service helpers module and moved the
associated tests.

rmkraus committed:
Merge remote-tracking branch 'balloob/dev' into automation-decorator

# Conflicts:
#	homeassistant/helpers/service.py
#	tests/helpers/test_service.py

rmkraus committed:
Fixed lint issue from merge

extract_entity_ids from the service helpers was overwriting the service
decorator with one of its attributes. This was fixed.

rmkraus committed:
Removed service decorator from event decorators

rmkraus commented:
Ok, I think we're ready for critical review.

balloob commented on homeassistant/helpers/event_decorators.py:
```diff
@@ -0,0 +1,81 @@
+HASS = None
+
+
+def _callback(action, *args, **kwargs):
```
Can you add `HASS` as the first argument to this method? That way we could potentially clean up the reference to HASS after bootstrap.

balloob commented on homeassistant/helpers/event_decorators.py:
```diff
@@ -0,0 +1,81 @@
+    def track_state_change_decorator(action):
+        """ Decorator to track state changes """
+        event.track_state_change(HASS, entity_ids,
+                                 functools.partial(_callback, action),
```
Actually, thinking about this, wouldn't the following yield exactly the same result:
```python
functools.partial(action, HASS)
```

balloob commented:
![giphy](https://cloud.githubusercontent.com/assets/1444314/12542783/8e067a06-c2dc-11e5-852a-bc4873c2efbc.gif)

Looks very good, I like it. One minor comment how I think we can optimize the callback.

Also, do you think we should remove the `HASS` globals after bootstrap is done?

rmkraus committed:
Removed decorator callback

The decorator callback was not actually necessary so it was removed and
replaced with a partial function instead.

rmkraus commented:
Replaced the decorator callback with a partial function.

I kept the HASS globals in the helper modules. I don't particularly like keeping them there either, but if someone were to write a custom component that imports another custom component after initialization, it could cause issues.
